### PR TITLE
[ISSUE #5049] Hook interface add execute order

### LINKED ISSUE #5049: Hook interface add execute order
#### Feature describe 

There have many Hook interface in RMQ, e.g: `SendMessageHook` `ConsumeMessageHook`  `PutMessageHook` etc. when its add to list,The order of execution is the order in which it is added to the list.  so we can add a order for Hook interface, and sorted hook which it is added to the same list

#### Solution

In hook interface add a method:

```java
    default int order(){
        return Integer.MIN_VALUE;
    }
```

when hook is added into list  then sorted by order number. Smaller values are executed first

####  Advantage

- Control over the order of execution is implemented through the interface, not through the order of joining the list

### PR DESCRIPTION
**Pull request by mxsm**
**Make sure set the target branch to `develop`**

## What is the purpose of the change

close #5049 

## Brief changelog

- Hook interface add execute order

## Verifying this change

XXXX

Follow this checklist to help us incorporate your contribution quickly and easily. Notice, `it would be helpful if you could finish the following 5 checklist(the last one is not necessary)before request the community to review your PR`.

- [x] Make sure there is a [Github issue](https://github.com/apache/rocketmq/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a Github issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue. 
- [x] Format the pull request title like `[ISSUE #123] Fix UnknownException when host config not exist`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test(over 80% coverage) to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/rocketmq/tree/master/test).
- [x] Run `mvn -B clean apache-rat:check findbugs:findbugs checkstyle:checkstyle` to make sure basic checks pass. Run `mvn clean install -DskipITs` to make sure unit-test pass. Run `mvn clean test-compile failsafe:integration-test`  to make sure integration-test pass.
- [ ] If this contribution is large, please file an [Apache Individual Contributor License Agreement](http://www.apache.org/licenses/#clas).

### CODE REVIEW

File: broker/src/test/java/org/apache/rocketmq/broker/mqtrace/SendMessageHookTest.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+        list.sort(Comparator.comparing(SendMessageHook::order));
+        SendMessageHook hook = list.get(0);
+        Assert.assertEquals("hook1", hook.hookName());
+        list.add(hook2);
+        hook = list.get(2);
+        Assert.assertEquals("hook2", hook.hookName());
+        list.sort(Comparator.comparing(SendMessageHook::order));
+        hook = list.get(2);
+        Assert.assertEquals("hook3", hook.hookName());
+
```
Comment (lizhiboo) [inline broker/src/test/java/org/apache/rocketmq/broker/mqtrace/SendMessageHookTest.java:110]: IMO, it's essential to add a new test case that use default order value works as before.

Comment (mxsm) [reply broker/src/test/java/org/apache/rocketmq/broker/mqtrace/SendMessageHookTest.java:110]: I got it, I will add a new test case that use default order value works as before.

**Comment by zhouxinyu** [comment]
Hi, could you please describe more information about your scenes? Why doesn't the added order meet your demand? 

And, if we really want to implement this feature, the netty handler pipeline mechanism could be your reference.

**Comment by mxsm** [comment]
> Hi, could you please describe more information about your scenes? Why doesn't the added order meet your demand?
> 
> And, if we really want to implement this feature, the netty handler pipeline mechanism could be your reference.

Hi @zhouxinyu 

Multiple hooks to accomplish one thing requires the orchestration of hooks. for example PutMessageHook, the first PutMessageHook to check message common properties, e.g: topic length. The others hook which in list do handle others thing. So I want to orchestrate the execution order by adding order. 
It is mainly used to solve the scenario of sequential Hook execution. The order should not be the order in which the list is added but can be controlled in some way

**Comment by ni-ze** [comment]
> 
IMO,  It is nature that executed in the order in which they were added. If a task need completed by two order hooks, combine them into one task is better.

**Review by echooymxq — COMMENTED** [review]
IMHO, It is unnecessary, or why not implement a `Ordered` interface? and Other hooks extend it.

**Comment by mxsm** [comment]
Hi @ni-ze  I got it. Thks  I will close this PR

**Pull request closed without merging**